zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -0,0 +1,1107 @@
+                };
+
+                // get the block height for the end of the current shard
+                let range_max = sapling_shard_end_stmt
+                    .query_row(named_params![":shard_index": max_idx], |row| {
+                        row.get::<_, Option<u32>>(0)
+                            .map(|opt| opt.map(BlockHeight::from))
+                    })
+                    .optional()?
+                    .flatten();
+
+                Ok::<Range<BlockHeight>, rusqlite::Error>(match (range_min, range_max) {
+                    (Some(start), Some(end)) => Range { start, end },
+                    (Some(start), None) => Range {
+                        start,
+                        end: range.end,
+                    },
+                    (None, Some(end)) => Range {
+                        start: range.start,
+                        end,
+                    },
+                    (None, None) => range.clone(),
+                })
+            })
+            .transpose()
+            .map_err(SqliteClientError::from)
+    }?;
+
+    let query_range = extended_range.clone().unwrap_or_else(|| range.clone());
+
+    let scanned = ScanRange::from_parts(range.clone(), ScanPriority::Scanned);
+    let extensions = if let Some(extended) = extended_range {
+        vec![
+            ScanRange::from_parts(extended.start..range.start, ScanPriority::FoundNote),
+            ScanRange::from_parts(range.end..extended.end, ScanPriority::FoundNote),
+        ]
+    } else {
+        vec![]
+    };
+
+    replace_queue_entries(
+        conn,
+        &query_range,
+        Some(scanned).into_iter().chain(extensions.into_iter()),
+    )?;
+
+    Ok(())
+}
+
+pub(crate) fn update_chain_tip<P: consensus::Parameters>(
+    conn: &rusqlite::Transaction<'_>,
+    params: &P,
+    new_tip: BlockHeight,
+) -> Result<(), SqliteClientError> {
+    // Read the maximum height from the shards table.
+    let shard_start_height = conn.query_row(
+        "SELECT MAX(subtree_end_height)
+        FROM sapling_tree_shards",
+        [],
+        |row| Ok(row.get::<_, Option<u32>>(0)?.map(BlockHeight::from)),
+    )?;
+
+    // Create a scanning range for the fragment of the last shard leading up to new tip.
+    // However, only do so if the start of the shard is at a stable height.
+    let shard_entry = shard_start_height
+        .filter(|h| h < &new_tip)
+        .map(|h| ScanRange::from_parts(h..new_tip, ScanPriority::ChainTip));
+
+    // Create scanning ranges to either validate potentially invalid blocks at the wallet's view
+    // of the chain tip,
+    let tip_entry = block_height_extrema(conn)?.map(|(_, prior_tip)| {
+        // If we don't have shard metadata, this means we're doing linear scanning, so create a
+        // scan range from the prior tip to the current tip with `Historic` priority.
+        if shard_entry.is_none() {
+            ScanRange::from_parts(prior_tip..new_tip, ScanPriority::Historic)
```

Comment:
If the caller first calls `WalletDb::put_sapling_subtree_roots` then this should never be `None`. However, in testing while doing so, I never saw a `ChainTip` range for a newly recovered-from-seed wallet; I just saw a single `Historic` range from wallet birthday to chain tip. I will investigate further.